hpx/parallel/segmented_algorithms/detail/scan.hpp

```diff
@@ -0,0 +1,792 @@
+
+                return hpx::util::unused;
+            }
+
+            template <typename ExPolicy, typename InIter,
+                typename OutIter, typename T, typename Op>
+            static typename util::detail::algorithm_result<ExPolicy>::type
+            parallel(ExPolicy && policy, InIter first,
+                InIter last, OutIter dest, T && init, Op && op)
+            {
+                typedef typename util::detail::algorithm_result<ExPolicy>::type
+                    result_type;
+
+                if(first == last)
+                    return util::detail::algorithm_result<ExPolicy>::get();
+
+
+                return hpx::util::void_guard<result_type>(),
+                    Algo().parallel(
+                        std::forward<ExPolicy>(policy), first, last, dest,
+                        std::forward<T>(init), std::forward<Op>(op));
+            }
+        };
+
+        template <typename SegIter, typename OutIter>
+        static bool is_segmented_the_same(SegIter first, SegIter last, OutIter dest,
+            std::false_type)
+        {
+            return false;
+        }
+
+        // check if two segmented iterators are partitioned the same
+        // partition size and id should be the same
+        template <typename SegIter, typename OutIter>
+        static bool is_segmented_the_same(SegIter first, SegIter last, OutIter dest,
+            std::true_type)
+        {
+            typedef hpx::traits::segmented_iterator_traits<SegIter> traits_in;
+            typedef typename traits_in::segment_iterator segment_iterator_in;
+            typedef typename traits_in::local_iterator local_iterator_type_in;
+
+            typedef hpx::traits::segmented_iterator_traits<OutIter> traits_out;
+            typedef typename traits_out::segment_iterator segment_iterator_out;
+            typedef typename traits_out::local_iterator local_iterator_type_out;
+
+            segment_iterator_in sit_in = traits_in::segment(first);
+            segment_iterator_in send_in = traits_in::segment(last);
+
+            segment_iterator_out sit_out = traits_out::segment(dest);
+
+            if (sit_in == send_in)
+            {
+                // all elements on the same partition
+                local_iterator_type_in beg_in = traits_in::local(first);
+                local_iterator_type_in end_in = traits_in::end(sit_in);
+
+                local_iterator_type_out beg_out = traits_out::local(dest);
+                local_iterator_type_out end_out = traits_out::end(sit_out);
+
+                if(beg_in != end_in) {
+                    std::uint64_t in_id = traits_in::get_id(sit_in).get_msb();
+                    std::uint64_t out_id = traits_in::get_id(sit_in).get_msb();
+
+                    std::size_t in_dist = std::distance(beg_in, end_in);
+                    std::size_t out_dist = std::distance(beg_out, end_out);
+
+                    if (in_id != out_id || in_dist != out_dist) {
```

Comment:
Minor nitpick: there is no need to invoke `std::distance` if the ids are different.